### PR TITLE
Support different types of key when communicating via grpc

### PR DESCRIPTION
**Pull request by xieyxclack**
As the title says. The kye types can be `int` or `string`.

### CODE REVIEW

File: federatedscope/core/message.py [hunk 1]
```diff
@@ -180,7 +189,11 @@ def build_msg_value(self, value):
         if isinstance(value, list) or isinstance(value, tuple):
             msg_value.list_msg.MergeFrom(self.create_by_type(value))
         elif isinstance(value, dict):
-            msg_value.dict_msg.MergeFrom(self.create_by_type(value))
+            if isinstance(list(value.keys())[0], str):
```
Comment (rayrayraykk) [inline federatedscope/core/message.py:192]: What if a `dict` contains string key and int key?

**Review by rayrayraykk — COMMENTED** [review]
Please see the inline comments, and the hotfix looks good to me.

File: federatedscope/core/communication.py [hunk 2]
```diff
@@ -84,8 +84,14 @@ def serve(self, max_workers, host, port, options):
         return server
 
     def add_neighbors(self, neighbor_id, address):
-        self.neighbors[neighbor_id] = '{}:{}'.format(address['host'],
-                                                     address['port'])
+        if isinstance(address, dict):
+            self.neighbors[neighbor_id] = '{}:{}'.format(
+                address['host'], address['port'])
+        elif isinstance(address, str):
```
Comment (joneswong) [inline federatedscope/core/communication.py:90]: We'd better have an assertion about the format of string

File: federatedscope/core/message.py (hunk 1 above)
Comment (xieyxclack) [reply federatedscope/core/message.py:192]: Such situation cannot be handled due to the implementation grpc. A workaround can be that we change all the keys into string and save their original types for recovery, which might bring additional cost.

Comment (rayrayraykk) [reply federatedscope/core/message.py:192]: IMO, we'd better have a format check with dict values instead of using the first key.

**Review by joneswong — APPROVED** [review]
approved

**Pull request merged**